File: selection-sort/selection-sort.rb
module SelectionSort
  extend self
  require 'wikipedia'

  def sort(array, options = :asc)
    #iterate over the given array
    for i in (0..array.length - 1)
      # iterate over the remaining elements
      for j in ((i + 1)..array.length - 1)
        # if the elements are out of order, switch them
        if array[j] < array[i] && options == :asc
          array[j], array[i] = array[i], array[j]
        elsif array[j] > array[i] && options == :desc
          array[j], array[i] = array[i], array[j]
        end
      end
    end
    array
  end

  def profile
    page = Wikipedia.find( 'Selection Sort' )
    puts page.summary.split("\n").map { |a| a.strip }.join(" ").gsub(/{.+}/, "")
  end

end

File: selection-sort/java/SelectionSort.java
public class SelectionSort {

  public enum Order { ASC, DESC }

  public static int[] sort( int[] array ) {
    return sort(array, Order.ASC);
  }

  public static int[] sort( int[] array, Order order ) {
    for (int i = 0; i <= array.length - 1; i++) {
      for (int j = i + 1; j <= array.length - 1; j++) {
        if (array[j] < array[i] && order == Order.ASC) {
          swap(array, i, j);
        } else if (array[j] > array[i] && order == Order.DESC) {
          swap(array, i, j);
        }
      }
    }

    return array;
  }

  public static int[] swap( int[] array, int index1, int index2 ) {
    int temp = array[index1];
    array[index1] = array[index2];
    array[index2] = temp;

    return array;
  }
}
